File: Exercise 1/HeapSort/HeapSort.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <sstream>
#include <chrono>

using namespace std;
using namespace chrono;  // Thư viện đo thời gian

void Heapify(vector<int>& arr, int n, int i) {
    int largest = i, left = 2 * i + 1, right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;
    if (right < n && arr[right] > arr[largest])
        largest = right;

    if (largest != i) {
        swap(arr[i], arr[largest]);
        Heapify(arr, n, largest);
    }
}

// Heap Sort
void HeapSort(vector<int>& arr) {
    int n = arr.size();

    for (int i = n / 2 - 1; i >= 0; i--)
        Heapify(arr, n, i);

    for (int i = n - 1; i > 0; i--) {
        swap(arr[0], arr[i]);
        Heapify(arr, i, 0);
    }
}

int main() {
    ifstream inputFile("input.txt");
    vector<vector<int>> arrs;
    string line;

    // Read Input
    while (getline(inputFile, line)) {
        int n = stoi(line), num;
        vector<int> arr;

        if (getline(inputFile, line)) {
            stringstream ss(line);
            while (ss >> num)
                arr.push_back(num);
        }

        arrs.push_back(arr);
    }

    inputFile.close();

    // Sort and Check Time
    for (size_t i = 0; i < arrs.size(); i++) {
        auto start = high_resolution_clock::now();
        HeapSort(arrs[i]);
        auto end = high_resolution_clock::now();

        duration<double> exec_time = end - start;

        cout << "Array " << i + 1 << " (Sorted in " << exec_time.count() << " seconds)" << endl;
    }

    return 0;
}

File: Sorting Report/HeapSort/HeapSort.py
from time import *

# Read Input
with open('input.txt', 'r') as file:
    lines = file.readlines()
    arrs = []
    for i in range(0, len(lines), 2):
        n = int(lines[i].strip())
        arrs.append(list(map(int, lines[i + 1].strip().split())))

# Heap Sort
def Heapify(arr : list, n : int, i : int):
    largest = i
    low = 2 * i + 1
    high = 2 * i + 2

    if low < n and arr[i] < arr[low]:
        largest = low
    if high < n and arr[largest] < arr[high]:
        largest = high

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # Swap

        Heapify(arr, n, largest)

def HeapSort(arr : list):
    n = len(arr)

    for i in range(n // 2, -1, -1):
        Heapify(arr, n, i)

    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # Swap

        Heapify(arr, i, 0)

# Sort and Check time
for arr in arrs:
    start_time = time()
    HeapSort(arr)
    finish_time = time()

    print(f"Array (Sorted in {finish_time - start_time:.6f} seconds)")
